File: Day10-19/12.rb
def to_mips(inst, initial_mappings={})
    register_map = 'abcd'.chars.each_with_object({}) do |e,h|
        h[e] = '$t' + (e.ord - 'a'.ord).to_s 
    end
    init = initial_mappings.map do |k, v|
        "li \t#{register_map[k]}, #{v}"
    end
    main = inst.each_with_index.map do |line, idx|
        line = line.split(' ')
        case line[0]
        when 'cpy'
            dst = register_map[line[2]]
            src = register_map.fetch(line[1], line[1].to_i)
            "l#{idx}: \tadd \t#{dst}, $0, #{src}"
        when 'inc'
            reg = register_map[line[1]]
            "l#{idx}: \tadd \t#{reg}, #{reg}, 1"
        when 'dec'
            reg = register_map[line[1]]
            "l#{idx}: \tsub \t#{reg}, #{reg}, 1"
        when 'jnz'
            reg = register_map.fetch(line[1], line[1].to_i)
            off = line[2].to_i
            "l#{idx}: \tbne \t$0, #{reg}, l#{idx + off}"
        end
    end
    text = ".text\nmain:\n"
    text += "#{init.join("\n")}\n"
    text += "#{main.join("\n")}\n"
    text += "\tmove \t$a0, #{register_map['a']}\n"
    text += "\tli \t$v0, 1\n"
    text += "\tsyscall\n"
    text
end


File.open('12.txt', 'r') do |f|
    aoc_asm = f.read.split("\n")
    File.open('12_1.s', 'w+') do |w|
        mips = to_mips(aoc_asm)
        w.write mips
        puts mips
    end
    File.open('12_2.s', 'w+') do |w|
        mips = to_mips(aoc_asm, 'c' => 1)
        w.write mips
        puts mips
    end
end

File: Day10-19/12.py

def simulate(inst, initial_values={}):
    registers = {x: 0 for x in 'abcd'}
    registers.update(initial_values)

    def get_val(i):
        if i in registers:
            return registers[i]
        return int(i)

    instructs = map(lambda x: x.split(), inst)
    pc = 0
    while pc < len(instructs):
        op = instructs[pc]
        if op[0] == 'cpy':
            registers[op[2]] = get_val(op[1])

        elif op[0] == 'inc':
            registers[op[1]] += 1
        elif op[0] == 'dec':
            registers[op[1]] -= 1
        elif op[0] == 'jnz':
            if get_val(op[1]) != 0:
                pc += get_val(op[2])
                continue
        else:
            print "No opcode for %s" % op[0]
            break
        pc += 1
    return registers

if __name__ == '__main__':
    with open('12.txt', 'r') as f:
        inst = f.readlines()
        print("Part 1: %s" % simulate(inst)['a'])
        print("Part 2: %s" % simulate(inst, {'c': 1})['a'])
